fix: player networkobject not added to client owned object list

NoelStephensUnity opened the pull request:
This PR primarily addresses the issue where player NetworkObjects were not being added to the NetworkClients OwnedObjects list.  However, it was also discovered that dynamically spawned NetworkObjects were not properly updating the NetworkClients  OwnedObjects list as it was only updating this list if the ChangeOwnership or RemoveOwnership messages were sent (but not spawn messages).
This update fixes both issues by creating a NetworkSpawnManager.OwnershipToObjectsTable that is updated for all clients on the server-side and only for the owned objects of the client on the client-side.

[MTT-1585](https://jira.unity3d.com/browse/MTT-1585)

## Changelog

### com.unity.netcode.gameobjects
- Fixed: Player objects and dynamically spawned NetworkObjects not being added to the client's owned objects list.

## Testing and Documentation
* Includes modified integration tests.
* Includes integration tests.

NoelStephensUnity committed:
fix

This fixes the issue with a player not being added to the client's owned objects list.  This also fixes the issue where dynamically spawned NetworkObjects were not getting added to the client's owned object list.
This fix migrates the table into NetworkSpawnManager, removes the OwnedObjects list from NetworkClient (replaces it with a method), and provides automatic assignment and de-assignment of ownership to the NetworkSpawnManager's OwnershipToObjectsTable.

NoelStephensUnity committed:
tests

Updates some of the ownership integration tests to the changes.

NoelStephensUnity committed:
Update and Style

Log message and exception additions.
White space removal.
Extra-LF removal.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-1585-player-networkobject-not-in-client-owned-list

NoelStephensUnity committed:
fix

removed extra '}' char.

NoelStephensUnity committed:
fix

Fix for server ownership invocation.

NoelStephensUnity committed:
style

whitespace

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-1585-player-networkobject-not-in-client-owned-list

0xFA11 committed:
Merge branch 'develop' into fix/mtt-1585-player-networkobject-not-in-client-owned-list

NoelStephensUnity committed:
Merge branch 'fix/mtt-1585-player-networkobject-not-in-client-owned-list' of https://github.com/Unity-Technologies/com.unity.netcode.gameobjects into fix/mtt-1585-player-networkobject-not-in-client-owned-list

NoelStephensUnity committed:
fix

Fixing minor issue with NetworkTransform having ownership applied before OnNetworkSpawn being invoked.

0xFA11 commented on com.unity.netcode.gameobjects/Runtime/Connection/NetworkClient.cs:
```diff
@@ -18,8 +19,18 @@ public class NetworkClient
+            {
+                return PlayerObject.NetworkManager.SpawnManager.GetClientOwnedObjects(ClientId);
+            }
+            return null;
```
it might be better to return an empty list instead of `null`

NoelStephensUnity committed:
test

This is the test for this PR

0xFA11 committed:
update `NetworkObjectOwnershipTests`

NoelStephensUnity committed:
fix

set previous owner whether removing or not.

0xFA11 committed:
minor stuff

NoelStephensUnity committed:
fix

Set the ownership before invoking lost or gained ownership methods.

NoelStephensUnity committed:
Merge branch 'fix/mtt-1585-player-networkobject-not-in-client-owned-list' of https://github.com/Unity-Technologies/com.unity.netcode.gameobjects into fix/mtt-1585-player-networkobject-not-in-client-owned-list

NoelStephensUnity committed:
style

NoelStephensUnity committed:
style

NoelStephensUnity committed:
fix and update

fixing some order of operations issues.
reverting back to OwnedObjects get.

NoelStephensUnity committed:
style

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-1585-player-networkobject-not-in-client-owned-list

NoelStephensUnity committed:
update

removed the internal OwnerClientId.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-1585-player-networkobject-not-in-client-owned-list

NoelStephensUnity committed:
update

Removing nullable client id as a parameter in spawn methods.

NoelStephensUnity committed:
test

Minor update to the ownership test where it double checks the update to NetworkClient when checking if a client's player is included in the owned objects list.

NoelStephensUnity committed:
Update CHANGELOG.md

NoelStephensUnity committed:
style

Minor clean up.

NoelStephensUnity committed:
fix

Assure that clients that are neither the current or new owner of the NetworkObject update their properties to reflect the new owner.

NoelStephensUnity committed:
test

Added a test to verify that switching ownership between several clients works.

NoelStephensUnity committed:
style

updating small comment for clarity

NoelStephensUnity committed:
style

updated comments for clarity.

NoelStephensUnity committed:
Merge branch 'develop' into fix/mtt-1585-player-networkobject-not-in-client-owned-list

0xFA11 committed:
minor update

netcode-ci-service committed:
Merge develop into fix/mtt-1585-player-networkobject-not-in-client-owned-list
